src/treelearner/monotone_constraints.hpp

```diff
@@ -62,6 +62,25 @@ class LeafConstraintsBase {
       const std::vector<SplitInfo>& best_split_per_leaf) = 0;
 
   inline static LeafConstraintsBase* Create(const Config* config, int num_leaves);
+
+  double ComputeMonotoneSplitGainPenalty(int leaf_index, double penalization,
+                                                double epsilon = 1e-10) {
+    int depth = tree->leaf_depth(leaf_index);
+    if (penalization >= depth + 1.) {
+      return epsilon;
+    }
+    if (penalization <= 1.) {
+      return 1. - penalization / pow(2., depth) + epsilon;
+    }
+    return 1. - pow(2, penalization - 1. - depth) + epsilon;
+  }
+
+  void ShareTreePointer(Tree* tree) {
```

Comment:
Const T*

Reply:
I am not sure what you mean by that, since I am copying the pointer, I don't understand how I can having constant. 

If I just write void ShareTreePointer(const Tree* tree), then I am getting errors:
invalid conversion from ‘const LightGBM::Tree*’ to ‘LightGBM::Tree*’

Reply:
Never mind, I didn't see the change above.